crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -0,0 +1,297 @@
+use revm_primitives::U256;
+
+use crate::alloc::vec;
+use crate::alloc::vec::Vec;
+use core::{
+    cmp::min,
+    fmt,
+    ops::{BitAnd, Not},
+};
+
+/// A sequential memory shared between calls, which uses
+/// a `Vec` for internal representation.
+/// A [SharedMemory] instance should always be obtained using
+/// the `new` static method to ensure memory safety.
+pub struct SharedMemory {
+    /// Shared buffer
+    data: Vec<u8>,
+    /// Memory checkpoints for each depth
+    checkpoints: Vec<usize>,
+    /// Raw pointer used for the portion of memory used
+    /// by the current context
+    current_slice: *mut [u8],
+    /// How much memory has been used in the current context
+    current_len: usize,
+    /// Amount of memory left for assignment
+    pub limit: u64,
+}
+
+impl fmt::Debug for SharedMemory {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SharedMemory")
+            .field(
+                "current_slice",
+                &crate::primitives::hex::encode(self.get_current_slice()),
+            )
+            .finish()
+    }
+}
+
+impl SharedMemory {
+    /// Allocate memory to be shared between calls.
+    /// Memory size is estimated using https://2π.com/22/eth-max-mem
+    /// which depends on transaction [gas_limit].
+    /// Maximum allocation size is 2^32 - 1 bytes;
+    pub fn new(gas_limit: u64) -> Self {
```

Comment:
This can call `Self::new_with_memory_limit(u32::MAX as usize)` (also `pow` doesn't get optimized well I think)